### PR TITLE
Bugfix/invalid sourcemaps whitespace

### PR DESCRIPTION
**Pull request by Hoten**
Fixes https://github.com/webpack/webpack/issues/7616

OriginalSource created a source map that failed validation (`sourcemap-validator`). The `pos` variable tracks the column number for a particular line, but doesn't get incremented when the current token is all whitespace. I believe that only happened for leading whitespace, as all (but possibly the last) token returned by `_splitCode` would contain some non-whitespace character. The effect was off-by-some-columns errors in the mappings, which manifested in the bootstrap template for webpack (it has leading whitespace).

To create a test, I had to expand the interface for ReplaceSource.replace to get `names` populated in the source map.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,7 +27,8 @@
     "istanbul": "^0.4.1",
     "js-beautify": "^1.5.10",
     "mocha": "^3.4.2",
-    "should": "^11.2.1"
+    "should": "^11.2.1",
```
Comment (evilebottnawi) [inline package.json]: Why we need `should` here?

Comment (Hoten) [reply package.json]: Can you elaborate? The tests use `should`.

**Comment by Hoten** [comment]
I suppose older versions of Node don't support default values. I'll replace that code.

File: lib/ReplaceSource.js [hunk 2]
```diff
@@ -18,16 +18,16 @@ class ReplaceSource extends Source {
 		this.replacements = [];
 	}
 
-	replace(start, end, newValue) {
+	replace(start, end, newValue, oldValue = undefined) {
```
Comment (Hoten) [inline lib/ReplaceSource.js]: I suppose older versions of Node don't support default values. I'll replace this code.

File: package.json (hunk 1 above)
Comment (evilebottnawi) [reply package.json]: @Hoten oh, sorry my mistake, all good :+1:

**Comment by evilebottnawi** [comment]
@Hoten CI test failed, need fix

**Comment by Hoten** [comment]
Pushed fixes

File: lib/ReplaceSource.js [hunk 3]
```diff
@@ -18,16 +18,16 @@ class ReplaceSource extends Source {
 		this.replacements = [];
 	}
 
-	replace(start, end, newValue) {
+	replace(start, end, newValue, oldValue) {
```
Comment (michael-ciniawsky) [inline lib/ReplaceSource.js]:
```
newValue => replacer || replacment
oldValue => replacee || replaced
```

(Nitpick)

Comment (evilebottnawi) [reply lib/ReplaceSource.js]: @michael-ciniawsky not related to this PR, here bug fix, not refactoring

**Comment by sokra** [comment]
Thanks

**Comment by adaniliuk** [comment]
@Hoten did this PR fix the original issue? webpack@4.20.0 uses webpack-sources@1.3.0, but the issue is still reproducible in https://github.com/kenhoff/webpack-sourcemap-testing when webpack version is updated.

**Comment by Hoten** [comment]
Yeah, I'm still seeing the same issue.

Side note, your test case isn't as intended. You've locked webpack to `4.12.2`, and may be using the globally installed version for compilation (prefer `npx webpack` over `webpack`).

**Comment by adaniliuk** [comment]
@Hoten just in case I've forked the repo and updated webpack version to the latest and changed the scripts https://github.com/adaniliuk/webpack-sourcemap-testing

There is the output:
```
$ webpack
Hash: 9bd50b65233535f2f320
Version: webpack 4.22.0
Time: 361ms
Built at: 2018-10-22 10:42:17
        Asset       Size  Chunks             Chunk Names
    bundle.js  984 bytes       0  [emitted]  main
bundle.js.map   4.52 KiB       0  [emitted]  main
Entrypoint main = bundle.js bundle.js.map
[0] ./src/index.js 46 bytes {0} [built]

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/concepts/mode/

$ node test-sourcemaps.js
/webpack-sourcemap-testing/node_modules/sourcemap-validator/index.js:146
          throw new Error(errMsg);
          ^
Error: Warning: mismatched names
Expected: installedModules || 'installedModules' || 'installedModules' || "installedModules" || "installedModules"
Got:  	var installedM ||  	var installedMod ||  	var installedMod ||  	var installedMod ||  	var installedMod
Original Line:  	var installedModules = {};
Mapping: 1:17→2:0 "installedModules" in webpack:///webpack/bootstrap
    at /webpack-sourcemap-testing/node_modules/sourcemap-validator/index.js:146:17
    at Array.forEach (<anonymous>)
    at SourceMapConsumer_eachMapping [as eachMapping] (/webpack-sourcemap-testing/node_modules/sourcemap-validator/node_modules/source-map/lib/source-map/source-map-consumer.js:570:10)
    at validate (/webpack-sourcemap-testing/node_modules/sourcemap-validator/index.js:83:12)
    at Object.<anonymous> (/webpack-sourcemap-testing/test-sourcemaps.js:6:1)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
```

**Comment by edmorley** [comment]
> did this PR fix the original issue? webpack@4.20.0 uses webpack-sources@1.3.0, but the issue is still reproducible

Would you mind opening a new issue so this doesn't get lost in the closed PR comments?

**Comment by adaniliuk** [comment]
@edmorley sure, was going to do that https://github.com/webpack/webpack/issues/8302
Thank you.

**Comment by adaniliuk** [comment]
@Hoten was this issue really solved to you?

**Pull request merged**